rules/consistent-function-scoping.js

```diff
@@ -0,0 +1,131 @@
+'use strict';
+const getDocsUrl = require('./utils/get-docs-url');
+
+const MESSAGE_ID_ARROW = 'ArrowFunctionExpression';
+const MESSAGE_ID_FUNCTION = 'FunctionDeclaration';
+
+function checkReferences(scope, parent, scopeManager) {
+	if (!scope) {
+		return false;
+	}
+
+	const {references} = scope;
+	if (!references || references.length === 0) {
+		return false;
+	}
+
+	const hit = references.some(reference => {
+		const variable = reference.resolved;
+
+		if (!variable) {
+			return false;
+		}
+
+		const hitReference = variable.references.some(reference => {
+			return parent === reference.from;
+		});
+
+		if (hitReference) {
+			return true;
+		}
+
+		const hitDefinitions = variable.defs.some(definition => {
+			const scope = scopeManager.acquire(definition.node);
+			return parent === scope;
+		});
+
+		if (hitDefinitions) {
+			return true;
+		}
+
+		return false;
+	});
+
+	if (hit) {
+		return true;
+	}
+
+	return scope.childScopes.some(scope => {
+		return checkReferences(scope, parent, scopeManager);
+	});
+}
+
+function checkNode(node, scopeManager) {
+	const scope = scopeManager.acquire(node);
+	if (!scope) {
+		return true;
+	}
+
+	let parentNode = node.parent;
+	if (!parentNode) {
+		return true;
+	}
+
+	// Skip over junk like the block statement inside of a function declaration
+	// or the various pieces of an arrow function.
+
+	if (parentNode.type === 'VariableDeclarator') {
+		parentNode = parentNode.parent;
+	}
+
+	if (parentNode.type === 'VariableDeclaration') {
+		parentNode = parentNode.parent;
+	}
+
+	if (parentNode.type === 'BlockStatement') {
```

Comment:
Hm, yeah that's a good test case. What should happen is that it will get reported because you can safely move `doBar` up one scope.

The way this works is:

1. Check for function / arrow function
2. Check immediate parent for a block statement and skip it
3. Starting at first un-skipped ancestor, recursively check all descendants for a reference to a variable in the ancestor scope
4. If none found, warn

In the test case you posted, it _should_ correctly mark the extra block statement as a pointless scope.

Reply:
I tried this out in ast explorer. It didn't fire the lint rule https://astexplorer.net/#/gist/92ef00cc145f6f92dec0268d64ca4a03/ea7ebb21a6ad6c218795719dfbaedd633e4e97dc

Reply:
Looks like `scopeManager.acquire` won't return a scope for the extra block statement. I'm not entirely sure why but I can probably fix it today or tomorrow.

Most likely the easiest fix is similar to what you recommended: Run a loop until we find a scope, hit the global scope or run out of ancestors.

Good catch. 👍

Reply:
@ankeetmaini Alright, I took at look at fixing this and it turns out it is non-trivial. Trying to add a loop to track back up more than one parent caused all sorts of problems with things like `for` loops, variables defined within the blocks, functions used in an assignment and nested blocks already at root.

@sindresorhus, @futpib how much do either of you care about about detecting things inside of empty block scopes?

Reply:
> how much do either of you care about about detecting things inside of empty block scopes?

I don't care much. I think I've only used an empty block a couple of times in the past years. Just note the limitation in the docs and if someone cares they can add support for it.